File: codes/2025/16-Third-ECNU-100-32-47 (C++)/src/frontend/AST/stmt.cc
#include "AST/stmt.h"
#include "AST/expr.h"
#include "AST/decl.h"
#include <ostream>
#include <sstream>

namespace aaac {
namespace frontend {

AssignStmt::~AssignStmt() { delete var; delete expr; }

DeclStmt::~DeclStmt() { delete decl; };

CompondStmt::~CompondStmt() { 
    for(Stmt *stmt : stmts) delete stmt;
}

IfStmt::~IfStmt() { delete cond; delete then; delete elsee; }

WhileStmt::~WhileStmt() { delete cond; delete body; }

ReturnStmt::~ReturnStmt() { delete ret; }

static void Indent(std::ostream &os, int d) {
    os << std::string(d,' ');
}

std::string Stmt::getAsString() const {
    std::ostringstream os;
    PrintTo(os, 0);
    return os.str();
}

void AssignStmt::PrintTo(std::ostream &os, int d) const {
    Indent(os, d);
    var->PrintTo(os, 0);
    os << " = ";
    expr->PrintTo(os, 0);
    // os << ";\n";
}

void DeclStmt::PrintTo(std::ostream &os, int d) const {
    decl->PrintTo(os, d);
}

void CompondStmt::PrintTo(std::ostream &os, int d) const {
    // Indent(os, d);
    // os << "{\n";
    for(Stmt *stmt : stmts) {
        bool isCompound = false;
        if(typeid(stmt) == typeid(CompondStmt)) {
            isCompound = true;
        }
        if(isCompound) {
            Indent(os, d);
            os << "{\n";
        }
        stmt->PrintTo(os, d + 4 * isCompound);
        if(isCompound) {
            Indent(os, d);
            os << "}";
        } else {
            if(typeid(*stmt) != typeid(WhileStmt) &&
                typeid(*stmt) != typeid(IfStmt)) {
                os << ";";
            }
        }
        os << "\n";
    }
    // os << "\n";
    // Indent(os,d);
    // os << "}\n";
} 

void IfStmt::PrintTo(std::ostream &os, int d) const {
    Indent(os, d);
    os << "if (";
    cond->PrintTo(os, 0);
    os << ") {\n";
    then->PrintTo(os, d + 4);
    if(typeid(*then) != typeid(CompondStmt)) {
        os << ";\n";
    }
    Indent(os, d);
    os << "}";
    if(elsee) {
        os << " else {\n";
        elsee->PrintTo(os, d + 4);
        if(typeid(*elsee) != typeid(CompondStmt)) {
            os << ";\n";
        }
        Indent(os, d);
        os << "}";
    }
    // os << "\n";
}

void WhileStmt::PrintTo(std::ostream &os, int d) const {
    Indent(os, d);
    os << "while (";
    cond->PrintTo(os, 0);
    os << ") {\n";
    body->PrintTo(os, d + 4);
    Indent(os, d);
    os << "}";
}

void ReturnStmt::PrintTo(std::ostream &os, int d) const {
    Indent(os, d);
    os << "return ";
    if(ret) {
        ret->PrintTo(os, 0);
    }
    // os << ";\n";
}

void BreakStmt::PrintTo(std::ostream &os, int d) const {
    Indent(os, d);
    os << "break";
}

void ContinueStmt::PrintTo(std::ostream &os, int d) const {
    Indent(os, d);
    os << "continue";
}

void EmptyStmt::PrintTo(std::ostream &os, int d) const {
    return;
}

} // namespace frontend
} // namespace aaac

File: codes/2025/22-Excellence-XJTU-100-22-40 (Java)/src/main/java/cn/edu/xjtu/sysy/ast/node/Stmt.java
package cn.edu.xjtu.sysy.ast.node;

import cn.edu.xjtu.sysy.symbol.SymbolTable;
import org.antlr.v4.runtime.Token;

import java.util.List;

/** Stmts */
public abstract sealed class Stmt extends Node {
    private boolean isDead = false;

    public Stmt(Token start, Token end) {
        super(start, end);
    }

    public void dead() {
        isDead = true;
    }

    public boolean isDead() {
        return isDead;
    }

    /** BlockStmt */
    public static final class Block extends Stmt {
        /** 语句块元素 */
        public List<Stmt> stmts;

        /** 符号表*/
        public SymbolTable.Local symbolTable;

        public Block(Token start, Token end, List<Stmt> stmts) {
            super(start, end);
            this.stmts = stmts;
        }
    }

    /** VarDefStmt */
    public static final class LocalVarDef extends Stmt {
        public List<Decl.VarDef> varDefs;

        public LocalVarDef(Token start, Token end, List<Decl.VarDef> varDefs) {
            super(start, end);
            this.varDefs = varDefs;
        }
    }

    /** AssignStmt assignableExp '=' exp ';' */
    public static final class Assign extends Stmt {
        /** 赋值表达式的左值 */
        public Expr.Assignable target;

        /** 右值 */
        public Expr value;

        public Assign(Token start, Token end, Expr.Assignable target, Expr value) {
            super(start, end);
            this.target = target;
            this.value = value;
        }
    }

    /** Return Stmts */
    public static final class Return extends Stmt {
        public Expr value;

        public Return(Token start, Token end, Expr value) {
            super(start, end);
            this.value = value;
        }
    }

    /** ExprStmt */
    public static final class ExprEval extends Stmt {
        public Expr expr;

        public ExprEval(Token start, Token end, Expr expr) {
            super(start, end);
            this.expr = expr;
        }
    }

    /** Stmts */
    public static final class If extends Stmt {
        public Expr cond;
        public Stmt thenStmt;
        public Stmt elseStmt;

        public If(Token start, Token end, Expr cond, Stmt thenStmt, Stmt elseStmt) {
            super(start, end);
            this.cond = cond;
            this.thenStmt = thenStmt;
            this.elseStmt = elseStmt;
        }
    }

    /** WhileStmt */
    public static final class While extends Stmt {
        public Expr cond;
        public Stmt body;

        public While(Token start, Token end, Expr cond, Stmt body) {
            super(start, end);
            this.cond = cond;
            this.body = body;
        }
    }

    /** ContinueStmt */
    public static final class Continue extends Stmt {
        public Continue(Token start, Token end) {
            super(start, end);
        }
    }

    /** BreakStmt */
    public static final class Break extends Stmt {
        public Break(Token start, Token end) {
            super(start, end);
        }
    }

    /** EmptyStmt */
    public static final class Empty extends Stmt {
        public Empty(Token start, Token end) {
            super(start, end);
        }

        public static final Stmt.Empty INSTANCE = new Stmt.Empty(null, null);
    }
}
